service/controller/resource/appversionlabel/create.go

```diff
@@ -2,9 +2,105 @@ package appversionlabel
 
 import (
 	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/giantswarm/apiextensions/v3/pkg/apis/application/v1alpha1"
+	"github.com/giantswarm/apiextensions/v3/pkg/label"
+	"github.com/giantswarm/microerror"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/giantswarm/cluster-apps-operator/pkg/project"
+	"github.com/giantswarm/cluster-apps-operator/service/controller/key"
+	"github.com/giantswarm/cluster-apps-operator/service/internal/releaseversion"
 )
 
+// EnsureCreated checks for optional apps and ensures the app-operator version
+// label has the correct value.
 func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
-	// TODO Add logic.
+	cr, err := key.ToCluster(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	var apps []*v1alpha1.App
+	{
+		r.logger.Debugf(ctx, "finding optional apps for workload cluster %#q", key.ClusterID(&cr))
+
+		o := metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("%s!=%s", label.ManagedBy, project.Name()),
+		}
+		list, err := r.g8sClient.ApplicationV1alpha1().Apps(key.ClusterID(&cr)).List(ctx, o)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		for _, item := range list.Items {
+			apps = append(apps, item.DeepCopy())
+		}
+
+		r.logger.Debugf(ctx, "found %d optional apps for workload cluster %#q", len(apps), key.ClusterID(&cr))
+	}
+
+	{
+		var updatedAppCount int
+
+		if len(apps) > 0 {
```

Comment:
I know this is copy & paste, but can we return early if we don't have any app?
```suggestion
		if len(apps) == 0 {
		  return nil
		}
```

Reply:
I'm fine with improving stuff as we copy. 👍 

This is nicer thx!